sympy/parsing/tests/test_latex_lark.py

```diff
@@ -385,9 +385,9 @@ def _log(a, b=E):
     (r"\overline{x + y}", conjugate(x + y)),
     (r"\overline{x} + \overline{y}", conjugate(x) + conjugate(y)),
     (r"\min(a, b)", Min(a, b)),
-    (r"\min(a, b, c - d, xy)", Min(a, b, c - d, x * y)),
+    (r"\min(a, b, c - x, xy)", Min(a, b, c - x, x * y)),
     (r"\max(a, b)", Max(a, b)),
-    (r"\max(a, b, c - d, xy)", Max(a, b, c - d, x * y)),
+    (r"\max(a, b, c - x, xy)", Max(a, b, c - x, x * y)),
```

Comment:
Why can't `d` be used as symbols here?

Reply:
If we disallow normal `d` and require all instances of `d` as a differential to be either `\text{d}` or `\mathrm{d}` then we can use `d` as a variable again. However, I don't think most people make a distinction between `d` as a differential and a variable in their LaTeX code.

Reply:
It is very common to use `d` as a variable name.

What about allowing `d` as a symbol and then having `dfrac` process into a derivative afterwards?

Reply:
I may suggest in other ways, to split start symbol into two expression cases

```
start: expression_with_derivative | expression_without_derivative
```

and defining `expression_with_derivative`, `expression_without_derivative` separately into their own recursive grammars such that
`expression_with_derivative` contains the derivatives or integrals, and `expression_without_derivative` does not.

I think that if users use `d` as polynomial variables, then users should deliberately avoid using derivatives at math, and vice versa. However, I'm not sure if this is easy before factoring out the recursive rules nicely.

> What about allowing d as a symbol and then having dfrac process into a derivative afterwards?

We can try that idea, but if we encouter some issues like 

- The expression `dx` parses to multiplication
- SymPy may sort the multiplication arguments

then we may rethink about the fix though.

Reply:
> The expression `dx` parses to multiplication

This is precisely the problem since the transformer is applied in a bottom-up fashion on the AST.

Reply:
It is probably better to have a two-pass process here. In the first pass `dx` becomes `Juxtapose(d, x)`. The second pass can decide to change something like `Dfrac(Symbol('d'), Juxtapose(Symbol('d'), Symbol('x'))` into a derivative or a fraction.

Reply:
I will try a few things and see if they stick, but I'm not sure how this two-pass approach would work. The Transformer runs only once. After that, we are supposed to have the desired output.

Reply:
The good news is that the integral feature is completely disjoint from the derivative feature. It looks to me like `\int \frac{dz}[z}` still parses as we want it to, even if we allow `d` as a symbol name.

Reply:
The transformer class for Lark ASTs has a very specific structure and is designed to handle a specific class of inputs. From my POV as a SymPy developer, I simply write the transformer class with the terminal or rule names as functions and do what I need to in the body, and Lark takes care of the parsing and transformer application.

I don't think this second pass wouldn't work because Lark flattens the tree as it applies the transformer class to the parse tree.

Reply:
> I cannot make this one-pass approach into a two-pass approach at will.

Lark does the first pass. Then you can do whatever you want with the object that Lark returns: that is when you do the second pass.

Reply:
Having everything in bottom-up logic, should make it more easier to implement things like evaluation controls in the future.

https://github.com/sympy/sympy/issues/25620

I think that evaluation control is more complicated if we also consider the case about top-down evaluation pass

Reply:
Fair enough.

If I was designing a parser like this I would make everything a two pass process from the outset. The first pass would just parse the latex and produce a symbolic representation of the latex itself (using expression heads that are all LaTeX-specific). The second pass would be a set of match/replacement rules that translate that into an expression. Then it is much easier for a user to control or customise the parsing by changing or adding their own match/replacement rules like
```python
rules = {
    LaTeX(r"\pi"): sympy.pi,
    LaTeX(r"\beta"): Symbol('beta'),
    ...
}
```

Reply:
I've made it so that `d` followed by another variable is parsed as a differential. This is not exactly what we want, but it's pretty darn close.

> However, I'm not sure if this is easy before factoring out the recursive rules nicely.

We can revisit the issue and put the final nail in the coffin once I have made a PR which splits the grammar into separate file.

Reply:
There is ISO standard 
https://tug.org/tugboat/tb18-1/tb54becc.pdf
that recommends typesetting upright `d` as differential.

Reply:
Have a look at the Mathematica parser. It has four steps and makes usage of lists as if they were Lisp S-expressions.

I suggest we go that way after Lark's transformer has been applied.

Reply:
> There is ISO standard
> https://tug.org/tugboat/tb18-1/tb54becc.pdf
> that suggests typesetting upright d as differential.

The problem is that most people either don't know about this convention or don't care. Take a look at [this highly popular Math StackExchange answer](https://math.stackexchange.com/a/565626/737093) for an example. All the `d`s are typeset in italics. However, the question put the `d` in the integral in an upright position.

Reply:
> The problem is that most people either don't know about this convention or don't care. Take a look at this highly popular Math StackExchange answer for an example.

I think that we can be loose about other iso standards, for easy things like not setting `i, j` or `e` as roman letters, because there is less technical difficulty implementing that.

However, for derivative, the technical difficulty of implementing parser can make a valid defense against some user requests about parser not being flexible though.

Reply:
> However, for derivative, the technical difficulty of implementing parser can make a valid defense against some user requests about parser not being flexible though.

True. I'd like to point out, however, that it works now even with a normal `d`. The only pathological case we have sacrificed is something like `ab + dt` but I'd say that that comes up very rarely in real applications.

Reply:
If possible, it would be good to make the parser as customizable as possible. Please do not make the current parser code too complicated in order to handle very specific rules. We should rather focus on making the parser customizable.

Reply:
The next PR will add the features that we discussed in our last meeting and add customizability to the parser.

> Please do not make the current parser code too complicated in order to handle very specific rules. 

I believe that your suggestion of multiple passes to parse `d` correctly would also add to the code complexity of the parser.

Reply:
Also, the reason I don't want to open multiple concurrent PRs is because merge conflicts are a pain. I already had to resolve merge conflicts for this PR, and I actually made a mistake which I then had to fix.